clock.go

```diff
@@ -233,204 +210,180 @@ func (m *Mock) Tick(d time.Duration) <-chan time.Time {
+type clockTimers []*mockTimer
 
-// clockTimers represents a list of sortable timers.
-type clockTimers []clockTimer
+func (a *clockTimers) Len() int { return len(*a) }
+func (a *clockTimers) Swap(i, j int) {
+	(*a)[i], (*a)[j] = (*a)[j], (*a)[i]
+	(*a)[i].index = i
+	(*a)[j].index = j
+}
+func (a *clockTimers) Less(i, j int) bool {
+	return (*a)[i].next.Before((*a)[j].next)
+}
 
-func (a clockTimers) Len() int           { return len(a) }
-func (a clockTimers) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
-func (a clockTimers) Less(i, j int) bool { return a[i].Next().Before(a[j].Next()) }
+func (a *clockTimers) Pop() any {
+	s := *a
+	item := s[len(s)-1]
+	s[len(s)-1] = nil
+	*a = s[:len(s)-1]
+	item.index = -1
+	return item
+}
 
-// Timer represents a single event.
-// The current time will be sent on C, unless the timer was created by AfterFunc.
-type Timer struct {
-	C       <-chan time.Time
-	c       chan time.Time
-	timer   *time.Timer // realtime impl, if set
-	next    time.Time   // next tick time
-	mock    *Mock       // mock clock, if set
-	fn      func()      // AfterFunc function, if set
-	stopped bool        // True if stopped, false if running
+func (a *clockTimers) Push(x any) {
+	t := x.(*mockTimer)
+	t.index = len(*a)
+	*a = append(*a, t)
 }
 
 // Stop turns off the ticker.
 func (t *Timer) Stop() bool {
-	if t.timer != nil {
-		return t.timer.Stop()
+	if t.realtime != nil {
+		return t.realtime.Stop()
 	}
-
-	t.mock.mu.Lock()
-	registered := !t.stopped
-	t.mock.removeClockTimer((*internalTimer)(t))
-	t.stopped = true
-	t.mock.mu.Unlock()
-	return registered
+	return t.mocktime.stop()
 }
 
 // Reset changes the expiry time of the timer
 func (t *Timer) Reset(d time.Duration) bool {
-	if t.timer != nil {
-		return t.timer.Reset(d)
-	}
-
-	t.mock.mu.Lock()
-	defer t.mock.mu.Unlock()
-
-	t.next = t.mock.now.Add(d)
-
-	registered := !t.stopped
-	if t.stopped {
-		t.mock.timers = append(t.mock.timers, (*internalTimer)(t))
+	if t.realtime != nil {
+		return t.realtime.Reset(d)
 	}
 
-	t.stopped = false
-	return registered
+	return t.mocktime.reset(d)
 }
 
-type internalTimer Timer
-
-func (t *internalTimer) Next() time.Time { return t.next }
-func (t *internalTimer) Tick(now time.Time) {
-	// a gosched() after ticking, to allow any consequences of the
-	// tick to complete
-	defer gosched()
-
-	t.mock.mu.Lock()
-	defer t.mock.mu.Unlock()
-
-	// Check if we're stopped, we could have been de-registered after taking the lock and before
-	// calling tick.
-	if t.stopped {
-		return
-	}
-	if t.fn != nil {
-		// defer function execution until the lock is released, and
-		defer func() { go t.fn() }()
-	} else {
-		// it's possible to reset the clock without draining the channel. Don't block in
-		// that case.
-		select {
-		case t.c <- now:
-		default:
-		}
-	}
-	t.stopped = true
-	t.mock.removeClockTimer((*internalTimer)(t))
+// Ticker is a mock implementation of [time.Ticker].
+type Ticker struct {
```

Comment:
Restructured to be smaller (when not using mocktime) and to avoid having to abstract over Ticker versus Timer in the mock logic.